File: cpp/src/turbineex1.cpp
/* 

  turbineex1.cpp

*/

#include "turbineex1.hpp"

TurbineEx1::TurbineEx1(umComponent dictComp)
{
    name = any_cast<const char *>(dictComp["name"]);
    ef = any_cast<double>(dictComp["ef"]);
    iPort = new Port(any_cast<mPort>(dictComp["iPort"]));
    oPort = new Port(any_cast<mPort>(dictComp["oPort"]));
    ePort = new Port(any_cast<mPort>(dictComp["ePort"]));

    portdict = {{"iPort", iPort},
                {"oPort", oPort},
                {"ePort", ePort}};
    energy = WORKEXTRACTED;
}

TurbineEx1::~TurbineEx1()
{
    delete iPort;
    delete oPort;
    delete ePort;
}

void TurbineEx1::state()
{
    if (ef == 1.0)
    {
        ePort->s = iPort->s;
        ePort->ps();
        oPort->s = iPort->s;
        oPort->ps();
    }
    else
    {
        double isoh = seups(ePort->p, iPort->s, 4);
        ePort->h = iPort->h - ef * (iPort->h - isoh);
        ePort->ph();
        isoh = seups(oPort->p, ePort->s, 4);
        oPort->h = ePort->h - ef * (ePort->h - isoh);
        oPort->ph();
    };
}

int TurbineEx1::balance()
{

    oPort->fdot = iPort->fdot - ePort->fdot;
    double ienergy = iPort->fdot * iPort->h;
    double oenergy = ePort->fdot * ePort->h + oPort->fdot * oPort->h;
    workExtracted = ienergy - oenergy;
    if (isnan(workExtracted))
    {
        return 0;
    }
    else
        return 1;
}

void TurbineEx1::setportaddress()
{
    if (iPort != portdict["iPort"])
        iPort = portdict["iPort"];
    if (oPort != portdict["oPort"])
        oPort = portdict["oPort"];
    if (ePort != portdict["ePort"])
        ePort = portdict["ePort"];
}

string TurbineEx1::resultstring()
{
    string result;
    result = "\n" + name;
    result += "\n" + Port::title;
    result += "\n" + iPort->resultstring();
    result += "\n" + oPort->resultstring();
    result += "\n" + ePort->resultstring();
    result += "\nworkExtracted(kJ/kg): " + to_string_with_precision<double>(workExtracted, 3) + "\n";
    return result;
}

File: python/components/turbineex1.py
"""
General Object-oriented Abstraction  of Rankine Cycle 

    TurbineEx1 class: 
       
        iPort inlet steam   
                 ┌────────┐
              ↓ ╱         │ 
               ┤          │
                ╲         │
                 └──┬─────┤
          ePort     ↓     ↓  oPort exhausted steam  
extracted steam     1    
  

json object example

     {
            "name": "Turbine1",
            "type": "TURBINEEX1",
            "ef": 0.85,
            "iPort":,
            "oPort": j,
            "ePort": k
      } 

 Author:Cheng Maohua  Email: cmh@seu.edu.cn

"""
from seuif97 import *
from .port import *


class TurbineEx1:

    energy = 'workExtracted'
    devtype = 'TURBINEEX1'

    def __init__(self, dictDev):
        self.name = dictDev['name']
        self.iPort = [Port(dictDev['iPort'])]
        self.oPort = [Port(dictDev['oPort'])]
        self.ePort = [Port(dictDev['ePort'])]

        self.ef = dictDev['ef']

        # map the name of port to the port obj
        self.portdict = {
            "iPort": self.iPort,
            "oPort": self.oPort,
            "ePort": self.ePort
        }

        self.workExtracted = 0
        self.WExtracted = 0

    def state(self):
        if self.ef == 1.0:
            self.ePort[0].s = self.iPort[0].s
            self.ePort[0].ps()
            self.oPort[0].s = self.iPort[0].s
            self.oPort[0].ps()
        else:
            isoh = ps2h(self.ePort[0].p, self.iPort[0].s)
            self.ePort[0].h = self.iPort[0].h - \
                self.ef * (self.iPort[0].h - isoh)
            self.ePort[0].ph()
            isoh = ps2h(self.oPort[0].p, self.ePort[0].s)
            self.oPort[0].h = self.ePort[0].h - \
                self.ef * (self.ePort[0].h - isoh)
            self.oPort[0].ph()

    def balance(self):
        """ mass and energy balance of the TurbineEx1
            work=ienergy - oenergy
        """
        self.oPort[0].fdot = self.iPort[0].fdot - self.ePort[0].fdot

        ienergy = self.iPort[0].fdot * self.iPort[0].h
        oenergy = self.ePort[0].fdot*self.ePort[0].h + \
            self.oPort[0].fdot*self.oPort[0].h
        self.workExtracted = ienergy - oenergy

    def sm_energy(self):
        """ mdot，get WExtracted """
        ienergy = self.iPort[0].mdot * self.iPort[0].h
        oenergy = self.ePort[0].mdot*self.ePort[0].h + \
            self.oPort[0].mdot*self.oPort[0].h
        self.WExtracted = ienergy - oenergy
        self.WExtracted /= (3600.0 * 1000.0)

    def __str__(self):
        result = '\n' + self.name
        result += '\n' + " PORT " + Port.title
        result += '\n' + " iPort " + self.iPort[0].__str__()
        result += '\n' + " oPort " + self.oPort[0].__str__()
        result += '\n' + " ePprt " + self.ePort[0].__str__()
        result += '\nworkExtracted(kJ/kg): \t{:>.2f}'.format(
            self.workExtracted)
        result += '\nWExtracted(MW): \t{:>.2f}'.format(self.WExtracted)
        return result
